### PR TITLE
Make API usage of FSharpExprPatterns possible with typed union fields

### PR DESCRIPTION
**Pull request by dawedawe**
Hey everyone,

FSC offers the `FSharpExprPatterns` Active Patterns API to easily traverse the TAST.
Unfortunately, the very common F# construct of a typed union field triggers an exception when traversed:
```fsharp
type T = Case1 of string
```
I hope the testcase in this PR helps to show the issue.

The reason is that DUs are processed by `FSharp.Compiler.AugmentWithHashCompare` to add the members `CompareTo`, `GetHashCode`, `Equals`

https://github.com/dotnet/fsharp/blob/91ff67b5f698f1929f75e65918e998a2df1c1858/src/Compiler/Checking/AugmentWithHashCompare.fs#L360

The `TDSwitch` that is constructed there has no default `DecisionTree`, it's `None`.
Later this is the reason that the exception is thrown at https://github.com/dotnet/fsharp/blob/91ff67b5f698f1929f75e65918e998a2df1c1858/src/Compiler/Symbols/Exprs.fs#L1269

The exception throwing code is quite old, so I was wondering if we could improve this today.
But I'm really unsure what a good handling should look like. So this PR has just my best guess (commented out) for a change in Exprs.fs to make traversal possible for code that contains a typed union field.

I'd be very grateful for any pointers how to improve the current situation.

### CODE REVIEW

**Comment by dawedawe** [comment]
So, uhm, anyone interested in this?
As this API is really nice otherwise, I'm pretty sure there are API users out there who could profit from an improvement.

File: src/Compiler/Symbols/Exprs.fs [hunk 1]
```diff
@@ -1264,9 +1264,12 @@ module FSharpExprConvert =
         match x with 
         | TDSwitch(inpExpr, csl, dfltOpt, m) -> 
             let acc = 
-                match dfltOpt with 
-                | Some d -> ConvDecisionTreePrim cenv env dtreeRetTy d 
-                | None -> wfail( "FSharp.Compiler.Service cannot yet return this kind of pattern match", m)
+                match dfltOpt, dtreeRetTy with
```
Comment (T-Gro) [inline src/Compiler/Symbols/Exprs.fs:1267]: Most of the times, we want to stripTypeEqns on a ttype - to arease aliases and units of measures.
What about here?

File: src/Compiler/Symbols/Exprs.fs [hunk 2]
```diff
@@ -1264,9 +1264,12 @@ module FSharpExprConvert =
         match x with 
         | TDSwitch(inpExpr, csl, dfltOpt, m) -> 
             let acc = 
-                match dfltOpt with 
-                | Some d -> ConvDecisionTreePrim cenv env dtreeRetTy d 
-                | None -> wfail( "FSharp.Compiler.Service cannot yet return this kind of pattern match", m)
+                match dfltOpt, dtreeRetTy with 
+                | Some d, _ -> ConvDecisionTreePrim cenv env dtreeRetTy d 
+                | None, TType.TType_app(tyconRef = _tyconRef; typeInstantiation = _typeInstantiation; flags = _flags)
```
Comment (T-Gro) [inline src/Compiler/Symbols/Exprs.fs:1269]: Underscores should be used at discarded values, not when the value is actually to be used.

File: src/Compiler/Symbols/Exprs.fs [hunk 3]
```diff
@@ -1264,9 +1264,12 @@ module FSharpExprConvert =
...
+                    when _tyconRef.CompiledName = "int" || _tyconRef.CompiledName = "bool" ->
```
Comment (T-Gro) [inline src/Compiler/Symbols/Exprs.fs:1270]: I must admit I do not see why we have to special case exactly int and bool (and nothing else)/
What is the reason?

Comment (dawedawe) [reply src/Compiler/Symbols/Exprs.fs]: I'm trying to limit the change of behaviour to a minimum while still getting rid of the exception for typed union fields.
And well, this is what seemed to be the minimal change to me.

`AugmentWithHashCompare.fs` adds `CompareTo` and `GetHashCode` which have type `int` and it adds `Equals` which has type `bool`.
These members are causing the `dfltOpt` to be `None` if the Union has typed fields, also see the description of this PR for the exact spots in the code.

But yeah, I also admit I'm not super confident that this is the correct way to do it.

File: src/Compiler/Symbols/Exprs.fs (hunk 2 above)
Comment (dawedawe) [reply src/Compiler/Symbols/Exprs.fs]: Sorry, fixed.

File: src/Compiler/Symbols/Exprs.fs (hunk 1 above)
Comment (dawedawe) [reply src/Compiler/Symbols/Exprs.fs]: Thanks, makes sense here, too.

File: src/Compiler/Symbols/Exprs.fs (hunk 3 above)
Comment (dawedawe) [reply src/Compiler/Symbols/Exprs.fs]: Mmh, I can't really think of a meaningful risk. So I'm okay with going without the limiting.

Comment (T-Gro) [reply src/Compiler/Symbols/Exprs.fs]: In that case, I would suggest:

- Do not limit if things should work anyway
- If limiting for the purpose of auto-generated methods (CompareTo,GetHashCode), limit just those explicitely.

**Review by psfinaki — APPROVED** [review]
So this can go I think? @T-Gro?

**Review by edgarfgp — APPROVED** [review]

**Comment by dawedawe** [comment]
Hey @T-Gro
I'm not quite sure if I understood your last comment correctly. Is this now in an acceptable state for you?

**Review by T-Gro — APPROVED** [review]

**Comment by knocte** [comment]
@T-Gro thanks for merging. BTW, just curious, if this was included in a commit from about 2 months ago, and the last release of FCS (43.8.101) was published 9 days ago, why is this fix not included in this last release?

**Comment by knocte** [comment]
@dawedawe do you know? ^

**Comment by vzarytovskii** [comment]
> @T-Gro thanks for merging. BTW, just curious, if this was included in a commit from about 2 months ago, and the last release of FCS (43.8.101) was published 9 days ago, why is this fix not included in this last release?

*.*.xx1 releases are point releases, and all additional features are not backported to those, only critical fixes.

**Comment by knocte** [comment]
> X.Y.xx1 releases are fix releases, and all additional features are not backported to those, only critical fixes.

Thanks for your answer Vlad. Do you know then which release number would then include this, and roughly when is it going to hit nuget.org?

**Comment by vzarytovskii** [comment]
> > X.Y.xx1 releases are fix releases, and all additional features are not backported to those, only critical fixes.
> 
> Thanks for your answer Vlad. Do you know then which release number would then include this, and roughly when is it going to hit nuget.org?

I think 8.0.200 should include it, you can test new versions here if you'd like to: https://dev.azure.com/dnceng/public/_artifacts/feed/dotnet8/NuGet/FSharp.Compiler.Service/overview/43.8.200-preview.24061.1

**Comment by vzarytovskii** [comment]
As for release date - I don't know, SDK publishes all of our packages, so it's tied to their release.

**Comment by knocte** [comment]
> I think 8.0.200 should include it

How does that version translate to FCS version numbers in nuget?

**Comment by vzarytovskii** [comment]
> > I think 8.0.200 should include it
> 
> How does that version translate to FCS version numbers in nuget?

It should be 43.8.200

**Comment by knocte** [comment]
> It should be 43.8.200

Cool thanks!

**Pull request merged**